### PR TITLE
CORE-5810 - Add helm chart that deploys the postgreSQL database used by app-simulator instances

### PR DESCRIPTION
**Pull request by dimosr**
## Changes
This adds a helm chart that can be used to deploy the postgreSQL database used by app-simulator instances to record data for sent/received messages used for testing analysis.

Most of the helm chart has been reused from https://github.com/corda/corda-dev-helm with only minor changes. One notable change is changing the init script, so that it creates the tables used by the app-simulator instances to record sent/received messages.

## Testing
After adding the repos and installing dependencies as explained in the README file, I deployed the helm chart using:
```
helm upgrade -i app-simulator-db -n draptis-app-simulator-db applications/tools/p2p-test/app-simulator/charts/app-simulator-db --create-namespace --render-subchart-notes --timeout 10m --wait
```
I tested it with:
```
helm test app-simulator-db -n draptis-app-simulator-db
```
And I also tested connecting to the DB with a UI (dbeaver)locally by using the following connection details after executing `sudo kubefwd svc -n draptis-app-simulator-db`:
* Host: app-simulator-db-postgresql
* Port: 5432
* Database: app_simulator
* Username: simulator
* Password: what was returned by `kubectl get secret --namespace draptis-app-simulator-db app-simulator-db-postgresql -o jsonpath="{.data.password}" | base64 --decode`

### CODE REVIEW

File: applications/tools/p2p-test/app-simulator/README.md [hunk 1]
```diff
@@ -226,4 +226,67 @@ group by time_window
...
+helm test "<RELEASE NAME>" -n "<RELEASE NAMESPACE>"
+```
+
+### Externalise services
+
+As an alternative to port-forwarding services for accessing them outside the cluster, a non-default `values-external.yaml` has been crafted - intended to decorate the existing `values.yaml` to expose postgreSQL via a NodePort service.
+
+It can be used by appending the following option to the `helm upgrade` command:
+```shell
+--values charts/app-simulator-db/values-external.yaml
```
Comment (williamvigorr3) [inline applications/tools/p2p-test/app-simulator/README.md:269]: Are you saying we should add this extra command when upgrading the `charts/corda-dev` helm chart?

File: applications/tools/p2p-test/app-simulator/README.md [hunk 2]
```diff
@@ -226,4 +226,67 @@ group by time_window
...
+```
+The node port used is 30200.
+
+## Maintaining
+
+As new value fields are added to the default `values.yaml`, doc strings should be included.
```
Comment (williamvigorr3) [inline applications/tools/p2p-test/app-simulator/README.md:275]: Presume the command doesn't actually modify this file. Just overrides the defaults?

File: applications/tools/p2p-test/app-simulator/charts/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+charts/app-simulator-db/charts/*
```
Comment (williamvigorr3) [inline applications/tools/p2p-test/app-simulator/charts/.gitignore]: So we ignore everything under `applications/tools/p2p-test/app-simulator/charts/charts/app-simulator-db/charts/*`? Is the double `charts` correct?

File: applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{{/*
+Selector labels
+*/}}
+{{- define "app-simulator-db.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "app-simulator-db.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end }}
+
+{{/*
+Create the name of the service account to use
```
Comment (williamvigorr3) [inline applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl:54]: Is there a reason we might want to use a separate service account. Presume we are using the default for now?

File: applications/tools/p2p-test/app-simulator/README.md (hunk 1 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/README.md:269]: I think there's been a misunderstanding here.. this chart is independend of the `corda-dev` one. The idea is you deploy a Corda cluster and then (if you want to use the app-simulator) you first deploy one instance of the DB with this chart and then a group of app-simulator on each cluster (with a chart yet to be created as part of the other stories).

Does this make sense or does the README contain something causing confusion?

File: applications/tools/p2p-test/app-simulator/README.md (hunk 2 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/README.md:275]: which command? What this really says is you change the defaults (which are the ones in `values.yaml`, follow this format and regenerate the docs README with the command below.

File: applications/tools/p2p-test/app-simulator/charts/.gitignore (hunk 3 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/charts/.gitignore]: hm, I think that's not right - it should be `app-simulator-db/charts/*`

Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/charts/.gitignore]: the answer to the other question is yes, I think this is a `charts` folder that is generated when building the chart.

**Comment by davidcurrie** [comment]
Is the PR missing a values.yaml for the new chart? Perhaps it might help answer the question that I have in my mind which is "does providing this chart actually provide any advantage over just pointing people at the postgres chart with a suitable set of overrides?".

**Comment by dimosr** [comment]
> Is the PR missing a values.yaml for the new chart? Perhaps it might help answer the question that I have in my mind which is "does providing this chart actually provide any advantage over just pointing people at the postgres chart with a suitable set of overrides?".

Sorry, I just realised that the `values.yaml` with the default values has not been picked up by git - not sure why. Let me look into it and I'll included it in the PR.

I preferred a separate chart for a couple of reasons:
* to prevent confusion: as conceptually, these are two different things - prerequisites for a Corda cluster deployment and prerequisites for an app-simulator deployment.
* to allow for independent evolution: currently, it seems that a subset of the dev-chart is sufficient as we deploy the postgresql, but this won't necessarily be the case in the future. Maybe we want to experiment with a different distributed SQL database at some point or generally move in some other direction that's not a good fit for the rest of the Corda cluster.
* The `values.yaml` (that I will now add in the PR) also contain the set of tables to be created. I'd prefer to that to be done out-of-the-box than requiring people to add it as an override on top of the `corda-dev-helm`.

**Comment by davidcurrie** [comment]
I wasn't suggesting conflating this with the `corda-dev` chart at all. I was suggesting that, given it appears to just provide a set of default values to be used with the `bitnami/postgresql` chart, why not just have users reference that chart directly and simply provide them with the override file needed for this specific scenario. (FWIW: I think the same is true for the `corda-dev` chart - its existence is pretty tenuous vs just deploying the kafka and postgresql charts independently.)

**Comment by davidcurrie** [comment]
I'd have to go looking for it but I think there may be some logic in the build pipeline at the moment that can't cope with there being multiple charts in the charts directory e.g. doing a `charts/*` and only expecting one result. That may need fixing up.

File: applications/tools/p2p-test/app-simulator/README.md (hunk 1 above)
Comment (williamvigorr3) [reply applications/tools/p2p-test/app-simulator/README.md:269]: Just read David's comment. I was mostly confused because this document doesn't contain `values.yml` anywhere in it (+ you forgot to add one in this PR). So I assumed that you were talking about the top level one. Should the `helm upgrade` command on line 253 have `--values values.yaml` in it? Maybe you don't need it.

It might be slightly clearer to not mention the `values.yaml` file at all.

... a non default `values-external.yaml` has been crafted - intended to override the default parameters to expose postgreSQL via a NodePort service.

File: applications/tools/p2p-test/app-simulator/README.md (hunk 2 above)
Comment (williamvigorr3) [reply applications/tools/p2p-test/app-simulator/README.md:275]: Again I think I was confused because you forgot to include the file.

**Comment by dimosr** [comment]
ok, `values.yaml` now checked in. It was being ignored (not sure where from), so I had to force add it.

> I wasn't suggesting conflating this with the `corda-dev` chart at all. I was suggesting that, given it appears to just provide a set of default values to be used with the `bitnami/postgresql` chart, why not just have users reference that chart directly and simply provide them with the override file needed for this specific scenario. (FWIW: I think the same is true for the `corda-dev` chart - its existence is pretty tenuous vs just deploying the kafka and postgresql charts independently.)

ah ok, misunderstanding on my side. Hmm, I guess there's sort of a similar argument, i.e. providing a slightly more out-of-the-box experience, so that people don't even have to specify a custom `values.yaml`. But, not so strong of an argument, so if you think there are strong reasons to avoid wrapping the bitnami one and using that directly instead (I'm not the expert on helm things..), happy to re-consider.

> I'd have to go looking for it but I think there may be some logic in the build pipeline at the moment that can't cope with there being multiple charts in the charts directory e.g. doing a `charts/*` and only expecting one result. That may need fixing up.

ah, ok - do you remember which part of the pipeline that might have been? Looking at the previous run, most of the steps (up to integration tests, which failed due to the expired certs issue that has now been resolved) succeeded. We can also see what happens with the latest one..

File: applications/tools/p2p-test/app-simulator/README.md (hunk 1 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/README.md:269]: > So I assumed that you were talking about the top level one. Should the `helm upgrade` command on line 253 have `--values values.yaml` in it? Maybe you don't need it.

yes, the chart is supposed to work without specifying a `values.yaml` file at all, the defaults should contain everything needed.

> It might be slightly clearer to not mention the `values.yaml` file at all.

ok, let me rephrase as you suggested if it's confusing.

File: applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl (hunk 4 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl:54]: yes, if I read that properly. If no custom service account is used in the values file, the `default` is used which reading [k8s docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/) seems to be a general default as well.

Not sure if we should be having another value as a default, since it seems to be something quite tied to the environment it's being deployed to, so the person that deploys would know better what they want.

Comment (williamvigorr3) [reply applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl:54]: Okay. The reason I ask is we don't do this here: [_helpers.tpl](https://github.com/corda/corda-runtime-os/blob/release/os/5.0/charts/corda/templates/_helpers.tpl)

File: applications/tools/p2p-test/app-simulator/README.md (hunk 1 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/README.md:269]: Changed that now, but I used "enriched" instead of "override" as it's not really overriding any of the default values atm.

File: applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl (hunk 4 above)
Comment (dimosr) [reply applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl:54]: it's the pre-reqs chart this is based on (so this [_helpers.tpl](https://github.com/corda/corda-dev-helm/blob/main/charts/corda-dev/templates/_helpers.tpl#L53-L62))

File: .jekyll-cache/Jekyll/Cache/Jekyll--Cache/b7/9606fb3afea5bd1609ed40b622142f1c98125abcfe89a76a661b0e8e343910 [hunk 5]
```diff
@@ -0,0 +1 @@
...th/:day/:title:output_ext"}}, "safe"=>false, "include"=>[".htaccess"], "exclude"=>[".sass-cache", ".jekyll-cache", "gemfiles", "Gemfile", "Gemfile.lock", "node_modules", "vendor/bundle/", "vendor/cache/", "vendor/gems/", "vendor/ruby/"], "keep_files"=>[".git", ".svn"], "encoding"=>"utf-8", "markdown_ext"=>"markdown,mkdown,mkdn,mkd,md", "strict_front_matter"=>false, "show_drafts"=>nil, "limit_posts"=>0, "future"=>false, "unpublished"=>false, "whitelist"=>[], "plugins"=>[], "markdown"=>"kramdown", "highlighter"=>"rouge", "lsi"=>false, "excerpt_separator"=>"\n\n", "incremental"=>false, "detach"=>false, "port"=>"4000", "host"=>"127.0.0.1", "baseurl"=>nil, "show_dir_listing"=>false, "permalink"=>"date", "paginate_path"=>"/page:num", "timezone"=>nil, "quiet"=>false, "verbose"=>false, "defaults"=>[], "liquid"=>{"error_mode"=>"warn", "strict_filters"=>false, "strict_variables"=>false}, "kramdown"=>{"auto_ids"=>true, "toc_levels"=>[1, 2, 3, 4, 5, 6], "entity_output"=>"as_char", "smart_quotes"=>"lsquo,rsquo,ldquo,rdquo", "input"=>"GFM", "hard_wrap"=>false, "guess_lang"=>true, "footnote_nr"=>1, "show_warnings"=>false}, "livereload_port"=>35729, "serving"=>true, "watch"=>true, "url"=>nil}:ET
```
Comment (williamvigorr3) [inline .jekyll-cache/Jekyll/Cache/Jekyll--Cache/b7/9606fb3afea5bd1609ed40b622142f1c98125abcfe89a76a661b0e8e343910]: Did you commit this by accident?

Comment (dimosr) [reply .jekyll-cache/Jekyll/Cache/Jekyll--Cache/b7/9606fb3afea5bd1609ed40b622142f1c98125abcfe89a76a661b0e8e343910]: yup, sorry about that - I'll revert it.

File: applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl (hunk 4 above)
Comment (davidcurrie) [reply applications/tools/p2p-test/app-simulator/charts/app-simulator-db/templates/_helpers.tpl:54]: Custom service accounts is still on the to-do list for the Corda Helm chart.

**Comment by davidcurrie** [comment]
> But, not so strong of an argument, so if you think there are strong reasons to avoid wrapping the bitnami one and using that directly instead (I'm not the expert on helm things..), happy to re-consider.

No - I don't really have a strong counterargument either other than that there's obviously a significant amount of boilerplate here when you're just overriding some values of another chart. Providing a custom chart would be more compelling if we were sharing a pre-built chart. As it is, having to pull the dependencies probably means there's more work for the end-user, not less.

> do you remember which part of the pipeline that might have been?

Searching for `charts/*` in https://github.com/corda/corda-shared-build-pipeline-steps/blob/5.0/vars/cordaPipeline.groovy shows a few places where I don't think it's going to do the right thing. Realistically, I'm not sure how generic these pipelines are meant to be. The lint should probably cover every chart but some of the version handling doesn't scale to multiple charts.

**Comment by dimosr** [comment]
> Providing a custom chart would be more compelling if we were sharing a pre-built chart. As it is, having to pull the dependencies probably means there's more work for the end-user, not less.

I see, my long-term plan was to possibly distribute a pre-built chart for this one even if that's purely for internal use, as we were planning to use these charts for load /stress testing. And we were considering adding jobs in Jenkins that would allow us to automate running these tests (even if just triggered manually).

> Searching for `charts/*` in https://github.com/corda/corda-shared-build-pipeline-steps/blob/5.0/vars/cordaPipeline.groovy shows a few places where I don't think it's going to do the right thing. Realistically, I'm not sure how generic these pipelines are meant to be. The lint should probably cover every chart but some of the version handling doesn't scale to multiple charts.

Given the build currently runs (probably not applying some things on the new chart, but at least not having issues with the Corda chart, which is the important bit), would we comfortable with merging this as-is and me following up on these things with @ronanbrowne to see how much we can generalise the build to apply these things globally on all charts (potentially with some toggles on custom things that don't apply everywhere)? So that we can at least make some progress, while working on that in parallel.

**Review by davidcurrie — APPROVED** [review]
Scratch all that about the build pipeline. I'd somehow got it into my head that this sat alongside the Corda Helm chart. It doesn't so none of the above applies. As I say, no religious objections to this being a separate chart so happy to approve.

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Comment by dimosr** [comment]
Merged this for now and I followed up separately with Ronan to see what we need to do to enable anything already in the pipeline (linting, publishing etc.) for these new charts as well.
